onnxmltools/utils/float16_converter.py

```diff
@@ -0,0 +1,113 @@
+                    if hasattr(n, 'attribute'):
+                        next_level.append(n.attribute)
+            # if q is model.graph.node.attribute, push q.g and q.graphs (GraphProto)
+            if hasattr(q, 'g'):
+                next_level.append(q.g)
+
+            if hasattr(q, 'graphs'):
+                for n in q.graphs:
+                    next_level.append(n)
+            # if q is graph, process graph.initializer(TensorProto), input and output (ValueInfoProto)
+            if hasattr(q, 'initializer'):
+                for n in q.initializer:
+                    if n.data_type == onnx_proto.TensorProto.FLOAT:
+                        n.data_type = onnx_proto.TensorProto.FLOAT16
+                        # convert float_data (float) to float16 and write to int32_data
+                        int_list = _npfloat16_to_int(np.float16(n.float_data))
+                        n.int32_data[:] = int_list
+                        n.float_data[:] = []
+                        # convert n.raw_data (bytes) to float
+                        float32_list = []
+                        for i in range(len(n.raw_data)//4):
+                            float32_list += unpack('f', n.raw_data[i*4:(i+1)*4])
+                        # convert float to float16
+                        int_list = _npfloat16_to_int(np.float16(float32_list))
+                        # convert float16 to bytes and write back to raw_data
+                        n.raw_data = b''
+                        for num in int_list:
+                            n.raw_data += num.to_bytes(2, byteorder='little', signed=False)
+            # graph.input (ValueInfoProto)
+            if hasattr(q, 'input'):
+                for n in q.input:
+                    # check if the input has type of ValueInfoProto (GraphProto) or string (NodeProto)
+                    if hasattr(n, 'type'):
+                         if n.type.tensor_type.elem_type == onnx_proto.TensorProto.FLOAT:
+                            n.type.tensor_type.elem_type = onnx_proto.TensorProto.FLOAT16
+            # graph.output (ValueInfoProto)
+            if hasattr(q, 'output'):
+                for n in q.output:
+                    # check if the output has type of ValueInfoProto (GraphProto) or string (NodeProto)
+                    if hasattr(n, 'type'):
+                        if n.type.tensor_type.elem_type == onnx_proto.TensorProto.FLOAT:
+                            n.type.tensor_type.elem_type = onnx_proto.TensorProto.FLOAT16
+            # if q is graph, process graph.value_info (ValueInfoProto)
+            if hasattr(q, 'value_info'):
+                for n in q.value_info:
+                    if n.type.tensor_type.elem_type == onnx_proto.TensorProto.FLOAT:
+                        n.type.tensor_type.elem_type = onnx_proto.TensorProto.FLOAT16
+            # if q is node.attribute, process node.attribute.t and node.attribute.tensors (TensorProto)
+            if hasattr(q, 't'):
```

Comment:
It's a bit dangerous. The field `t` is not a very meaningful name so adding type checking before accessing attributes might be necessary.